src/main/java/org/wildfly/security/auth/provider/FileSystemSecurityRealm.java

```diff
@@ -377,41 +371,42 @@ public void create() throws RealmUnavailableException {
             }
         }
 
-        public void setCredential(final Object credential) throws RealmUnavailableException {
-            Assert.checkNotNullParam("credential", credential);
+        public void deleteCredential(final String credentialName) throws RealmUnavailableException {
+            Assert.checkNotNullParam("credentialName", credentialName);
             final LoadedIdentity loadedIdentity = loadIdentity(false, false);
             if (loadedIdentity == null) {
                 throw ElytronMessages.log.fileSystemRealmNotFound(name);
             }
-            List<Object> credentials = loadedIdentity.getCredentials();
-            removeCredentialOfTheSameType(credentials, credential);
-            credentials.add(credential);
+
+            Map<String, Object> credentials = loadedIdentity.getCredentials();
+            credentials.remove(credentialName);
 
             final LoadedIdentity newIdentity = new LoadedIdentity(name, credentials, loadedIdentity.getAttributes());
             replaceIdentity(newIdentity);
         }
 
-        private void removeCredentialOfTheSameType(List<Object> credentials, Object pattern) {
-            for (Iterator iterator = credentials.iterator(); iterator.hasNext();) {
-                Object credential = iterator.next();
-                if (credential instanceof Key && pattern instanceof Key) { // keys comparison
-                    if (((Key)credential).getAlgorithm().equals(((Key)pattern).getAlgorithm())) {
-                        iterator.remove();
-                    }
-                } else { // non-keys comparison
-                    if(credential.getClass() == pattern.getClass()) {
-                        iterator.remove();
-                    }
-                }
+        public void setCredential(final String credentialName, final Object credential) throws RealmUnavailableException {
+            Assert.checkNotNullParam("credentialName", credentialName);
+            Assert.checkNotNullParam("credential", credential);
+            final LoadedIdentity loadedIdentity = loadIdentity(false, false);
+            if (loadedIdentity == null) {
+                throw ElytronMessages.log.fileSystemRealmNotFound(name);
             }
+
+            Map<String, Object> credentials = loadedIdentity.getCredentials();
+            credentials.put(credentialName, credential);
```

Comment:
I think this can cause ```UnsupportedOperationException``` given that the map returned from ```loadedIdentity.getCredentials()``` may be an ```Collections.emptyMap()```.